File: SuperSimpleStockMarket_Python/com/sssm/stocks/stock.py
from enum import Enum

class StockException(Exception):
    pass

class StockType(Enum):
    COMMON      = "Common Stock"
    PREFERRED   = "Preferred Stock"
    
class Stock(object):
    """ Represents a stock on the market
        stock_symbol identifies a stock and is used for equality check
        
        @author Lucky (LakshmiNarayana Ambarkar)
    """
    
    def __init__(self, stock_symbol, stock_type, last_dividend, fixed_dividend_percent, par_value):
        """ Create a stock object with symbol, type and par value.
            For simplicity, setting last_divident fixed_dividend_percent through the initializer method. 
        """
        self.__stock_symbol             = stock_symbol
        self.__stock_type               = stock_type
        self.__last_dividend            = last_dividend
        self.__fixed_dividend_percent   = fixed_dividend_percent
        self.__par_value                = par_value

        if stock_type == StockType.PREFERRED:
            self.__last_dividend = (fixed_dividend_percent/100) * par_value;
            
    def __str__(self):
        # nicely represent the stock on console.
        return self.__stock_type.value + " " + self.__stock_symbol
    
    def __eq__(self, other_stock):
        return self.__stock_symbol == other_stock.get_stock_symbol()

    def get_stock_symbol(self):
        return self.__stock_symbol
    
    def set_stock_type(self, stockType):
        self.__stock_type  = stockType
    
    def get_stock_type(self):
        return self.__stock_type
    
    def get_last_dividend(self):
        return self.__last_dividend
    
    def get_fixed_dividend_percent(self):
        return self.__fixed_dividend_percent
    
    def get_par_value(self):
        return self.__par_value
    
    def calculate_dividend_yield(self, stock_price):
        if (stock_price <= 0):
            raise StockException("Invalid price for " + str(self))
        
        # for PREFERRED stocks, lastDividend = parValue X fixedDividend/100.
        return self.get_last_dividend() / stock_price;
    
    def calculate_PE_ratio(self, stock_price):
        if (stock_price <= 0):
            raise StockException("Invalid price for " + str(self))
        
        if (self.__last_dividend == 0):
            raise StockException("Dividend never issued for " + str(self))
        
        return stock_price / self.__last_dividend

File: SuperSimpleStockMarket_Java/src/com/sssm/stocks/Stock.java
package com.sssm.stocks;

import com.sssm.stocks.exception.StockException;

/**
 * @author Lucky (LakshmiNarayana Ambarkar)
 * 
 * Represents a stock on the market
 * stockSymbol identifies a stock and is used for equality check
 */
public class Stock {
	public static enum StockType {
		COMMON,
		PREFERRED
	}
	
	private final String stockSymbol;
	private StockType stockType;
	private float lastDividend;
	private float fixedDividendPercent;
	private float parValue;
	
	public Stock(String symbol, StockType type, float lastDividend, float fixedDividendPercent, float parValue) {
		// for simplicity, setting all instance variables through constructor.
		this.stockSymbol = symbol;
		this.stockType = type;
		this.lastDividend = lastDividend;
		this.fixedDividendPercent = fixedDividendPercent;
		this.parValue = parValue;
		
		if (type == StockType.PREFERRED) {
			this.lastDividend = (fixedDividendPercent/100) * parValue;
		}
	}
	
	public String getStockSymbol() {
		return stockSymbol;
	}

	public StockType getStockType() {
		return stockType;
	}

	public void setStockType(StockType stockType) {
		this.stockType = stockType;
	}

	public float getLastDividend() {
		return lastDividend;
	}

	public void setLastDividend(float lastDividend) throws StockException {
		if (parValue <= 0f) {
			throw new StockException("Invalid dividend for " + this);
		}
		this.lastDividend = lastDividend;
	}

	public float getFixedDividendPercent() {
		return fixedDividendPercent;
	}

	public void setFixedDividendPercent(float fixedDividendPercent) throws StockException{
		if (this.stockType == StockType.PREFERRED && fixedDividendPercent <= 0f) {
			throw new StockException("Invalid fixed dividend for " + this);
		}
		this.fixedDividendPercent = fixedDividendPercent;
	}

	public float getParValue() {
		return parValue;
	}

	public void setParValue(float parValue) throws StockException {
		if (parValue <= 0f) {
			throw new StockException("Invalid par value for " + this);
		}
		this.parValue = parValue;
	}

	@Override
	public String toString() {
		// nicely represent the stock on console.
		return this.stockType.name() + " stock " + this.stockSymbol;
	}
	
	@Override
	public boolean equals(Object otherStock) {
		if (otherStock == this) { return true; }
		if (otherStock == null || this.getClass() != otherStock.getClass()) { return false; }
		
		return this.stockSymbol.equals(((Stock)otherStock).stockSymbol);
	}
	
	@Override
	public int hashCode() {
		return this.stockSymbol.hashCode();
	}
	
	public float calculateDividendYield(float stockPrice) throws StockException {
		if (stockPrice <= 0) {
			throw new StockException("Invalid price for " + this);
		}
		
		// for PREFERRED stocks, lastDividend = parValue X fixedDividend/100.
		return this.lastDividend / stockPrice;
	}
	
	public float calculatePERatio(float stockPrice) throws StockException {
		if (stockPrice <= 0) {
			throw new StockException("Invalid price for " + this);
		}
		
		if (this.lastDividend == 0) {
			throw new StockException("Dividend never issued for " + this);
		}
		
		return stockPrice / this.lastDividend;
	}
}
